File: CSCI382/python/InsertionSort.py
# Sorts a subarray A[p..r] in quadratic time. Usually faster than recursive sorting algorithms
#   for small input size. "Small" depends on the machine but is usually between 7 and 50 elements.
# Runtime: O(n^2)

def InsertionSort(A, p, r):
    for j in range(p+1, r+1):
        key = A[j]
        i = j-1
        while i>=p and A[i]>key:
            A[i+1] = A[i]
            i -= 1
        A[i+1] = key

"""
#test it out
from sys import argv
A = [int(x) for x in argv[1:]]
InsertionSort(A, 0, len(A)-1)
print A
"""

File: CSCI382/c++/InsertionSort.cpp
#include <iostream>
using namespace std;

void insertion_sort(int A[], int n){
    int i;
    int key;
    for (int j=0; j<n; j++) {
        key=A[j];
        i=j-1;
        while (i>=0 && A[i]>key) {
            A[i+1]=A[i];
            i--;
        }
        A[i+1]=key;
    }
}

int main() {
    int A[] = {2,1,4,3,7,5,9,10,8,6};
    cout << A[0] << "\n";
    insertion_sort(A,2);
    cout << A[0] << "\n";
}
